File: backend/spec/factories/project.rb
FactoryBot.define do
  factory :project do
    project_developer

    sequence(:name) do |n|
      "Project #{n}"
    end

    status { "published" }
    development_stage { "scaling-up" }
    category { "forestry-and-agroforestry" }
    sdgs { [1, 4, 5] }
    target_groups { %w[urban-populations indigenous-peoples] }
    impact_areas { %w[pollutants-reduction carbon-emission-reduction] }

    verified { false }

    estimated_duration_in_months { 13 }

    country
    municipality
    department

    sequence(:description) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 4)
    end
    sequence(:problem) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 4)
    end
    sequence(:solution) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 4)
    end
    sequence(:sustainability) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 4)
    end
    sequence(:replicability) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 4)
    end
    sequence(:progress_impact_tracking) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 4)
    end
    sequence(:expected_impact) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 4)
    end

    sequence(:positive_financial_returns) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 2)
    end
    sequence(:last_year_sales_revenue) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 1)
    end

    climate_change_risks_identified { true }

    sequence(:limate_change_risks_details) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 2)
    end

    looking_for_funding { true }
    instrument_types { %w[grant loan] }
    ticket_size { "scaling" }
    sequence(:funding_plan) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Lorem.paragraph(sentence_count: 4)
    end

    received_funding { true }
    received_funding_amount_usd { 3000 }
    sequence(:received_funding_investor) do |n|
      Faker::Config.random = Random.new(n)
      Faker::Company.name
    end

    language { "en" }
    geometry { {type: "Polygon", coordinates: [[[-7.256596802202454, 39.42149705239956], [-5.510236906689473, 40.02677860935444], [-4.408874148363334, 39.419930008870296], [-5.205992932301829, 38.67169482742881], [-6.402685625872827, 38.39721372248553], [-7.059852379049463, 38.94985978831832], [-7.256596802202454, 39.42149705239956]]]} }

    trait :with_involved_project_developers do
      involved_project_developers { create_list(:project_developer, 2) }
      involved_project_developer_not_listed { true }
    end

    trait :with_project_images do
      project_images { [build(:project_image, project: nil), build(:project_image, project: nil)] }
    end

    trait :draft do
      status { "draft" }
    end

    trait :published do
      status { "published" }
    end

    trait :with_impacts do
      impact_calculated { true }
      project_biodiversity_impact { 0.1 }
      project_climate_impact { 0.1 }
      project_water_impact { 0.1 }
      project_community_impact { 0.1 }
      project_total_impact { 0.1 }
      municipality_biodiversity_impact { 0.1 }
      municipality_climate_impact { 0.1 }
      municipality_water_impact { 0.1 }
      municipality_community_impact { 0.1 }
      municipality_total_impact { 0.1 }
      hydrobasin_biodiversity_impact { 0.1 }
      hydrobasin_climate_impact { 0.1 }
      hydrobasin_water_impact { 0.1 }
      hydrobasin_community_impact { 0.1 }
      hydrobasin_total_impact { 0.1 }
      priority_landscape_biodiversity_impact { 0.1 }
      priority_landscape_climate_impact { 0.1 }
      priority_landscape_water_impact { 0.1 }
      priority_landscape_community_impact { 0.1 }
      priority_landscape_total_impact { 0.1 }
    end

    trait :with_demands do
      project_demands_calculated { true }
      municipality_demands_calculated { true }
      hydrobasin_demands_calculated { true }
      priority_landscape_demands_calculated { true }
      project_biodiversity_demand { 0.1 }
      project_climate_demand { 0.1 }
      project_water_demand { 0.1 }
      project_community_demand { 0.1 }
      municipality_biodiversity_demand { 0.1 }
      municipality_climate_demand { 0.1 }
      municipality_water_demand { 0.1 }
      municipality_community_demand { 0.1 }
      hydrobasin_biodiversity_demand { 0.1 }
      hydrobasin_climate_demand { 0.1 }
      hydrobasin_water_demand { 0.1 }
      hydrobasin_community_demand { 0.1 }
      priority_landscape_biodiversity_demand { 0.1 }
      priority_landscape_climate_demand { 0.1 }
      priority_landscape_water_demand { 0.1 }
      priority_landscape_community_demand { 0.1 }
    end
  end
end

File: frontend/types/project.ts
import { ValidGeometryType } from 'containers/forms/geometry/types';
import { ProjectGalleryImageType } from 'containers/forms/project-gallery/project-gallery-image/types';

import {
  DevelopmentStages,
  Languages,
  TicketSizes,
  ProjectStatus,
  ImpactAreas,
  Impacts,
} from 'enums';

import { Locations } from './locations';

/** Project images on responses */
export type ProjectImageType = {
  id?: string;
  cover: boolean;
  file: {
    small: string;
    medium: string;
    original: string;
  };
};

/** Common Project types */
export type ProjectBase = {
  category: string;
  country: Locations;
  description: string;
  development_stage: DevelopmentStages;
  estimated_duration_in_months: number;
  expected_impact: string;
  geometry: ValidGeometryType;
  impact_areas: string[];
  instrument_types: string[];
  involved_project_developer_not_listed: boolean;
  looking_for_funding: boolean;
  funding_plan: string;
  name: string;
  problem: string;
  progress_impact_tracking: string;
  received_funding: boolean;
  received_funding_amount_usd?: number;
  received_funding_investor?: string;
  relevant_links?: string;
  climate_change_risks_identified: boolean;
  climate_change_risks_details?: string;
  positive_financial_returns?: string;
  last_year_sales_revenue?: number;
  replicability: string;
  sdgs: number[];
  solution: string;
  sustainability: string;
  target_groups: string[];
  ticket_size?: TicketSizes;
  language: Languages;
  municipality: Locations;
  department: Locations;
  project_images: ProjectImageType[];
  trusted?: boolean;
  favourite?: boolean;
  status?: ProjectStatus;
  project_developer?: any; // Cannot use ProjectDeveloperType because linting will complain about circular references
  involved_project_developers?: any[]; // Cannot use ProjectDeveloperType because linting will complain about circular references
  /**
   * Whether the logged in investor has marked the project as funded. `null` if the user is not
   * logged in or not an investor.
   **/
  funded?: boolean;
};

/** Project impact properties  */
export type ProjectImpacts = {
  impact_calculated: boolean;
  municipality_biodiversity_impact?: number;
  municipality_climate_impact?: number;
  municipality_water_impact?: number;
  municipality_community_impact?: number;
  municipality_total_impact?: number;
  hydrobasin_biodiversity_impact?: number;
  hydrobasin_climate_impact?: number;
  hydrobasin_water_impact?: number;
  hydrobasin_community_impact?: number;
  hydrobasin_total_impact?: number;
  priority_landscape_biodiversity_impact?: number;
  priority_landscape_climate_impact?: number;
  priority_landscape_water_impact?: number;
  priority_landscape_community_impact?: number;
  priority_landscape_total_impact?: number;
};

/** Project entity structure, received on get requests */
export type Project = ProjectBase &
  ProjectImpacts & {
    // Properties especific form Project(s) responses
    id: string;
    latitude: number;
    longitude: number;
    project_images: ProjectImageType[];
    slug: string;
    trusted?: boolean;
    type: 'project';
    priority_landscape: Locations;
    created_at: string;
    updated_at: string;
  };

/** Project Form inputs */
export type ProjectForm = ProjectBase & {
  country_id: string;
  department_id: string;
  involved_project_developer_ids: string[];
  municipality_id: string;
  project_images_attributes: ProjectImageGallery[];

  // Not part of the payload
  involved_project_developer: number;
  project_gallery?: FileList;
};

/** Project images for creation/edition */
export type ProjectImagesAttributes = {
  file?: string;
  cover: boolean;
  id?: string;
  _destroy?: boolean;
};

/** Project creation/edition payload */
export type ProjectCreationPayload = Omit<
  ProjectForm,
  'involved_project_developer' | 'project_gallery' | 'slug' | 'project_images_attributes'
> & {
  project_images_attributes: ProjectImagesAttributes[];
};

export type ProjectUpdatePayload = ProjectCreationPayload & {
  id: string;
};

export type ProjectImageGallery = ProjectImagesAttributes & ProjectGalleryImageType;

export type ProjectsMapGeojson = {
  type: 'FeatureCollection';
  features: {
    type: 'Feature';
    id: string;
    geometry: {
      type: 'Point';
      coordinates: number[];
    };
    properties: {
      id: string;
      type: string;
      trusted: boolean;
      category: string;
    };
  }[];
};

export type ProjectsMap = {
  id: string;
  type: string;
  trusted: boolean;
  category: string;
  latitude: number;
  longitude: number;
};

export type ProjectMapParams = {
  'filter[category]'?: string;
  'filter[sdg]'?: number;
  'filter[instrument_type]'?: string;
  'filter[ticket_size]'?: string;
  //'filter[only_verified]'?: string; VERIFICATION FILTERS: HIDDEN
  'filter[full_text]'?: string;
};

export type ProjectImpactScores = {
  biodiversity: number | null;
  climate: number | null;
  water: number | null;
  community: number | null;
  total: number | null;
};

export type ProjectImpactAreasScores = {
  municipality: ProjectImpactScores;
  hydrobasin: ProjectImpactScores;
  priority_landscape: ProjectImpactScores;
};
